Sources/Datadog/Core/Upload/DataUploadStatus.swift

```diff
@@ -134,11 +107,37 @@ private let ignoredNSURLErrorCodes = Set([
     NSURLErrorCannotConnectToHost, // -1004
 ])
 
-/// Looks at the `networkError` and produces error for internal Telemetry feature if anything is going wrong.
-private func createTelemetryErrorIfNeeded(for networkError: Error) -> (message: String, error: Error?)? {
-    let nsError = networkError as NSError
-    guard nsError.domain == NSURLErrorDomain, !ignoredNSURLErrorCodes.contains(nsError.code) else {
-        return nil
+extension DataUploadError {
+    init?(status code: Int) {
+        guard let responseStatusCode = HTTPResponseStatusCode(rawValue: code) else {
+            // If status code is unexpected, do not produce an error for internal Telemetry - otherwise monitoring may
+            // become too verbose for old installations if we introduce a new status code in the API.
+            return nil
+        }
+
+        switch responseStatusCode {
+        case .accepted, .forbidden:
+            // These codes mean either success or the user configuration mistake - do not produce error.
+            return nil
+        case .unauthorized:
+            self = .unauthorized
+        case .internalServerError, .serviceUnavailable:
+            // These codes mean Datadog service issue - do not produce SDK error as this is already monitored by other means.
+            return nil
+        case .badRequest, .payloadTooLarge, .tooManyRequests, .requestTimeout:
+            // These codes mean that something wrong is happening either in the SDK or on the server - produce an error.
+            self = .httpError(statusCode: code)
+        case .unexpected:
+            return nil
+        }
```

Comment:
We could also do:
```swift
switch uploadStatus.error {
case let .httpError(statusCode: statusCode) where statusCode == 401:
    userLogger.error("⚠️ The client token you provided seems to be invalid.")
case let .httpError(statusCode: statusCode):
    self.telemetry?.error("Data upload finished with status code: \(statusCode)")
case let .networkError(error: error):
    self.telemetry?.error("Data upload finished with error", error: error)
case .none: break
}
```
and get rid of the `.unauthorized` error case: more flexibility but less clarity.